### ISSUE TITLE
HTTP 599 error on large file sources with new tornado default

### ISSUE THREAD

**Issue by bradthurber** [saltstack/salt#24048]
Using latest ___develop___ branch (post 2015.5) will cause ```HTTP 599 ContentLength too long error``` with large files (>100MB?). For example - this command with a large file (160MB) will cause the error.:

salt '*develop*' cp.get_url https://s3.amazonaws.com/karpractice/jboss-eap-6.4.0.zip /tmp/blah

Traced back to /salt/utils/http.py line 
```
...
382       result = HTTPClient().fetch(
...
```

This appears to be due to new use of tornado as default http client in http.query and the default  tornado limits for max_buffer_size (100MB?). As a workaround, adding the max_buffer_size argument will allow a larger file to download. An example that allows for 300MB:

```
...
382       result = HTTPClient(max_buffer_size=1024*1024*300).fetch(
...
```

It should be noted that this not only impacts cp.get_url, but modules that use cp.get_url - such as archive.extracted

Perhaps we need to be able to configure a download file size max now? I'm not sure how this impacts memory usage.

**Comment by jfindlay** [comment]
@bradthurber, thanks for the report.

**Comment by jfindlay** [comment]
@bradthurber, does this happen when you are setting `transport: tcp` in the master/minion config?  I cannot reproduce this error on normal zeromq transport.  (Salt uses tornado in places other than the tcp transport, which you probably already know.)

**Comment by jfindlay** [comment]
I have attempted to reproduce this error on 2015.8.0rc3 with the following results:

### successful `cp.get_url`
```console
# time salt jmoney-node cp.get_url https://kernel.org/pub/linux/kernel/v4.x/linux-4.1.4.tar.xz /tmp/linux-4.1.4.tar.xz
jmoney-node:
    /tmp/linux-4.1.4.tar.xz

real    0m7.118s
user    0m1.016s
sys     0m0.072s
```
```
# master log
[DEBUG   ] Sending event - data = {'_stamp': '2015-08-04T19:12:51.528727', 'minions': ['jmoney-node']}
[DEBUG   ] Sending event - data = {'tgt_type': 'glob', 'jid': '20150804131251528418', 'tgt': 'jmoney-node', '_stamp': '2015-08-04T19:12:51.529175', 'user': 'root', 'arg': ['https://kernel.org/pub/linux/kernel/v4.x/linux-4.1.4.tar.xz', '/tmp/linux-4.1.4.tar.xz'], 'fun': 'cp.get_url', 'minions': ['jmoney-node']}
[INFO    ] User root Published command cp.get_url with jid 20150804131251528418
[DEBUG   ] Published command details {'tgt_type': 'glob', 'jid': '20150804131251528418', 'tgt': 'jmoney-node', 'ret': '', 'user': 'root', 'arg': ['https://kernel.org/pub/linux/kernel/v4.x/linux-4.1.4.tar.xz', '/tmp/linux-4.1.4.tar.xz'], 'fun': 'cp.get_url'}
[DEBUG   ] Sending event - data = {'_stamp': '2015-08-04T19:12:56.537495', 'minions': ['jmoney-node']}
[DEBUG   ] Sending event - data = {'tgt_type': 'glob', 'jid': '20150804131256537232', 'tgt': 'jmoney-node', '_stamp': '2015-08-04T19:12:56.537836', 'user': 'root', 'arg': ['20150804131251528418'], 'fun': 'saltutil.find_job', 'minions': ['jmoney-node']}
[INFO    ] User root Published command saltutil.find_job with jid 20150804131256537232
[DEBUG   ] Published command details {'tgt_type': 'glob', 'jid': '20150804131256537232', 'tgt': 'jmoney-node', 'ret': '', 'user': 'root', 'arg': ['20150804131251528418'], 'fun': 'saltutil.find_job'}
[INFO    ] Got return from jmoney-node for job 20150804131256537232
[DEBUG   ] Sending event - data = {'fun_args': ['20150804131251528418'], 'jid': '20150804131256537232', 'return': {'tgt_type': 'glob', 'jid': '20150804131251528418', 'tgt': 'jmoney-node', 'pid': 7979, 'ret': '', 'user': 'root', 'arg': ['https://kernel.org/pub/linux/kernel/v4.x/linux-4.1.4.tar.xz', '/tmp/linux-4.1.4.tar.xz'], 'fun': 'cp.get_url'}, 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2015-08-04T19:12:56.628711', 'fun': 'saltutil.find_job', 'id': 'jmoney-node'}
[INFO    ] Got return from jmoney-node for job 20150804131251528418
[DEBUG   ] Sending event - data = {'fun_args': ['https://kernel.org/pub/linux/kernel/v4.x/linux-4.1.4.tar.xz', '/tmp/linux-4.1.4.tar.xz'], 'jid': '20150804131251528418', 'return': '/tmp/linux-4.1.4.tar.xz', 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2015-08-04T19:12:57.544845', 'fun': 'cp.get_url', 'id': 'jmoney-node'}
```
```
# minion log
[INFO    ] User root Executing command cp.get_url with jid 20150804131251528418
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150804131251528418', 'tgt': 'jmoney-node', 'ret': '', 'user': 'root', 'arg': ['https://kernel.org/pub/linux/kernel/v4.x/linux-4.1.4.tar.xz', '/tmp/linux-4.1.4.tar.xz'], 'fun': 'cp.get_url'}
[INFO    ] Starting a new job with PID 7979
[DEBUG   ] LazyLoaded cp.get_url
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506', 'aes')
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506')
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] file_roots is /srv/salt
[DEBUG   ] Requesting URL https://kernel.org/pub/linux/kernel/v4.x/linux-4.1.4.tar.xz using GET method
[INFO    ] User root Executing command saltutil.find_job with jid 20150804131256537232
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150804131256537232', 'tgt': 'jmoney-node', 'ret': '', 'user': 'root', 'arg': ['20150804131251528418'], 'fun': 'saltutil.find_job'}
[INFO    ] Starting a new job with PID 7986
[DEBUG   ] LazyLoaded saltutil.find_job
[INFO    ] Returning information for job: 20150804131256537232
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506', 'aes')
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506')
[INFO    ] Returning information for job: 20150804131251528418
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506', 'aes')
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506')
```

### unsuccessful `cp.get_url`
https://mirrors.kernel.org/ubuntu-releases/15.04/ reports this ubuntu ISO at 1 GiB.
```console
# time salt jmoney-node cp.get_url http://mirrors.kernel.org/ubuntu-releases/15.04/ubuntu-15.04-desktop-amd64.iso /tmp/ubuntu.iso
jmoney-node:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1045, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.7/site-packages/salt/modules/cp.py", line 309, in get_url
        return __context__['cp.fileclient'].get_url(path, dest, False, saltenv)
      File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 605, in get_url
        raise MinionError('Error: {0}'.format(query['error']))
    MinionError: Error: HTTP 599: Connection closed

real    0m1.306s
user    0m0.924s
sys     0m0.088s
```
```
# master log
[DEBUG   ] Sending event - data = {'_stamp': '2015-08-04T19:15:56.084479', 'minions': ['jmoney-node']}
[DEBUG   ] Sending event - data = {'tgt_type': 'glob', 'jid': '20150804131556084218', 'tgt': 'jmoney-node', '_stamp': '2015-08-04T19:15:56.084805', 'user': 'root', 'arg': ['http://mirrors.kernel.org/ubuntu-releases/15.04/ubuntu-15.04-desktop-amd64.iso', '/tmp/ubuntu.iso'], 'fun': 'cp.get_url', 'minions': ['jmoney-node']}
[INFO    ] User root Published command cp.get_url with jid 20150804131556084218
[DEBUG   ] Published command details {'tgt_type': 'glob', 'jid': '20150804131556084218', 'tgt': 'jmoney-node', 'ret': '', 'user': 'root', 'arg': ['http://mirrors.kernel.org/ubuntu-releases/15.04/ubuntu-15.04-desktop-amd64.iso', '/tmp/ubuntu.iso'], 'fun': 'cp.get_url'}
[INFO    ] Authentication request from jmoney-node
[INFO    ] Authentication accepted from jmoney-node
[INFO    ] Got return from jmoney-node for job 20150804131556084218
[DEBUG   ] Sending event - data = {'fun_args': ['http://mirrors.kernel.org/ubuntu-releases/15.04/ubuntu-15.04-desktop-amd64.iso', '/tmp/ubuntu.iso'], 'jid': '20150804131556084218', 'return': 'The minion function caused an exception: Traceback (most recent call last):\n  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1045, in _thread_return\n    return_data = func(*args, **kwargs)\n  File "/usr/lib/python2.7/site-packages/salt/modules/cp.py", line 309, in get_url\n    return __context__[\'cp.fileclient\'].get_url(path, dest, False, saltenv)\n  File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 605, in get_url\n    raise MinionError(\'Error: {0}\'.format(query[\'error\']))\nMinionError: Error: HTTP 599: Connection closed\n', 'success': False, 'cmd': '_return', '_stamp': '2015-08-04T19:15:56.340693', 'fun': 'cp.get_url', 'id': 'jmoney-node', 'out': 'nested'}
[DEBUG   ] Sending event - data = {'id': 'jmoney-node', '_stamp': '2015-08-04T19:15:56.376780', 'result': True, 'pub': '-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA1ApzQJ6DgEwXbwtXq8qP\nHh3374hhV6y20PF7qnDO5d5t6STcRXQxItbsgSbOArltXs4/ABomv2xQ1gLwJUAr\n+gHt4ZRsCLc6IfvddfnFTp0h+TLkKO58kiPbCfrM5GRJKlNuaJKuvz42CwDUGMLw\nCdy8cRLm1NwVpY0W72wQgETlLP8FDV5x5TvZMccCJXJuXn7yJSVTopseh8EuNcso\nndYNpA9oHq+bfE2Nq8imNLQZN7dCCiMutxPuilL8akYkREooTvoXO35OQATtlr5d\nlv2eq7voh4nvrOF5rBo+zmR+8LfAwGMAdQ974L+bqPas3RkCB+CumKOoehrCfz6F\n5QIDAQAB\n-----END PUBLIC KEY-----', 'act': 'accept'}
[DEBUG   ] Sending event - data = {'_stamp': '2015-08-04T19:15:56.402035', 'pretag': None, 'cmd': '_minion_event', 'tag': '_salt_error', 'data': {'message': 'The minion function caused an exception', 'args': ['The minion function caused an exception'], '_stamp': '2015-08-04T19:16:30.861495'}, 'id': 'jmoney-node'}
[ERROR   ] Received minion error from [jmoney-node]: The minion function caused an exception
```
```python-traceback
# minion log
[INFO    ] User root Executing command cp.get_url with jid 20150804131556084218
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150804131556084218', 'tgt': 'jmoney-node', 'ret': '', 'user': 'root', 'arg': ['http://mirrors.kernel.org/ubuntu-releases/15.04/ubuntu-15.04-desktop-amd64.iso', '/tmp/ubuntu.iso'], 'fun': 'cp.get_url'}
[INFO    ] Starting a new job with PID 8193
[DEBUG   ] LazyLoaded cp.get_url
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506', 'aes')
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506')
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] file_roots is /srv/salt
[DEBUG   ] Requesting URL http://mirrors.kernel.org/ubuntu-releases/15.04/ubuntu-15.04-desktop-amd64.iso using GET method
[INFO    ] Malformed HTTP message from %s: %s
[WARNING ] The minion function caused an exception
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1045, in _thread_return
    return_data = func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/modules/cp.py", line 309, in get_url
    return __context__['cp.fileclient'].get_url(path, dest, False, saltenv)
  File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 605, in get_url
    raise MinionError('Error: {0}'.format(query['error']))
MinionError: Error: HTTP 599: Connection closed
[DEBUG   ] SaltEvent PUB socket URI: ipc:///var/run/salt/minion/minion_event_c6d60dd2ab_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: ipc:///var/run/salt/minion/minion_event_c6d60dd2ab_pull.ipc
[DEBUG   ] Sending event - data = {'message': u'The minion function caused an exception', 'args': ('The minion function caused an exception',), '_stamp': '2015-08-04T19:16:30.861495'}
[DEBUG   ] Handling event "_salt_error\n\n\x83\xa7message\xda\x00'The minion function caused an exception\xa4args\x91\xda\x00'The minion function caused an exception\xa6_stamp\xba2015-08-04T19:16:30.861495"
[DEBUG   ] Forwarding salt error event tag=_salt_error
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506', 'aes')
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506')
[DEBUG   ] Failed to authenticate message
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506', 'clear')
[INFO    ] Returning information for job: 20150804131556084218
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506', 'aes')
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'jmoney-node', 'tcp://23.253.230.82:4506')
```

**Comment by DmitryKuzmenko** [comment]
The problem is in the following.
```cp.get_url``` uses ```tornado.httpclient``` that have 2 implementations: ```simple_httpclient``` that is used by default and ```curl_httpclient```.
Simple client keeps all data in RAM that is the reason it has default size limit for security.

I see 2 possible ways:
1. Adjust ```simple_httpclient``` size limit (and make it configurable through in config files)
2. Use curl that is more powerful and fast but in this way we'll add a new dependency to salt (pycurl).

@basepi what's your opinion? Maybe you'd like to ping somebody else?

BTW, the issue could be easily reproduced without salt:
```
$ python -m tornado.httpclient --logging=debug http://mirrors.kernel.org/ubuntu-releases/15.04/ubuntu-15.04-desktop-amd64.iso
[I 150806 13:41:13 http1connection:253] Malformed HTTP message from None: Content-Length too long
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/lib/python2.7/site-packages/tornado/httpclient.py", line 656, in <module>
    main()
  File "/usr/lib/python2.7/site-packages/tornado/httpclient.py", line 642, in main
    validate_cert=options.validate_cert,
  File "/usr/lib/python2.7/site-packages/tornado/httpclient.py", line 102, in fetch
    self._async_client.fetch, request, **kwargs))
  File "/usr/lib/python2.7/site-packages/tornado/ioloop.py", line 445, in run_sync
    return future_cell[0].result()
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
tornado.httpclient.HTTPError: HTTP 599: Connection closed
```

**Comment by basepi** [comment]
Thanks for the detective work here, @DmitryKuzmenko. I haven't yet decided which direction we're going to go. @techhat and I were talking about maybe trying to use the streaming_callback functionality as shown [here](http://codingrelic.geekhold.com/2011/10/tornado-httpclient-chunked-downloads.html) and [here](http://stackoverflow.com/questions/20018684/tornado-streaming-http-response-as-asynchttpclient-receives-chunks). The callbacks complicate things, though. He's going to take a look this afternoon.

I will note switching to one of the other backends works. So we could potentially add a fallback. We don't dep on requests anymore, but perhaps we could just fallback to urllib.

**Comment by DmitryKuzmenko** [comment]
@basepi Thank you!

As from docs streaming_callback is supported by both client implementations so we could choose any.

About simple vs curl, just googled around and talked with some guys here. There are two opposite opinions:
1. Curl is much better because it works, simple client is buggy, has problems with SSL and so on.
2. Simple client just works and enough for fetching files, curl adds extra dependency on a system-wide installed native library of different version on different distros and not all versions work good.

So as for me it looks better to implement chunked download with ```streaming_callback``` test it with large files different protocols using simple client. If everything would work good, keep simple client. (Probably add a config option allowing to choose)

I can do all the work, let me know if I should do this.

**Comment by basepi** [comment]
I think that sounds like a good idea, @DmitryKuzmenko. Please proceed with the chunked download plan.

I think it makes sense to do the chunked download if `stream=True`.

**Comment by DmitryKuzmenko** [comment]
PR #26212 fixes the problem. The fix contains the following:
1. Use chunked fetch
2. Specify big (100GB) body size limit
3. Specify big (1 hour) timeout

TODO:
1. Make body size configurable.
2. Make something with timeout. Either one or another of the following:
2.1 The best way is to reset timeout watchdog if fetch is in progress
2.2 Just make timeout configurable too.

**Comment by DmitryKuzmenko** [comment]
Found no way to don't kill fetch by request timeout. Made minion configuration option for it. So PR #26250 implements points 1 and 2.2 from the previous comment's TODO section.

**Comment by cachedout** [comment]
Resolved via #26252
